sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -30,18 +30,43 @@ import org.apache.spark.sql.types.{DataType, IntegerType}
  *  - Intra-partition ordering of data: In this case the distribution describes guarantees made
  *    about how tuples are distributed within a single partition.
  */
-sealed trait Distribution
+sealed trait Distribution {
+  /**
+   * The required number of partitions for this distribution. If it's None, then any number of
+   * partitions is allowed for this distribution.
+   */
+  def requiredNumPartitions: Option[Int]
+
+  /**
+   * Creates a default partitioning for this distribution, which can satisfy this distribution while
+   * matching the given number of partitions.
+   */
+  def createPartitioning(numPartitions: Int): Partitioning
+}
 
 /**
  * Represents a distribution where no promises are made about co-location of data.
  */
-case object UnspecifiedDistribution extends Distribution
+case object UnspecifiedDistribution extends Distribution {
+  override def requiredNumPartitions: Option[Int] = None
+
+  override def createPartitioning(numPartitions: Int): Partitioning = {
+    throw new IllegalStateException("UnspecifiedDistribution does not have default partitioning.")
+  }
+}
 
 /**
  * Represents a distribution that only has a single partition and all tuples of the dataset
  * are co-located.
  */
-case object AllTuples extends Distribution
+case object AllTuples extends Distribution {
```

Comment:
`SingleNodeDistribution`?

Reply:
So I spoke with @sameeragarwal about this a little.  The whole point here was to have a logical / physical separation.  `AllTuples` could be `SingleNode` or it could be `Broadcast`.  All the operation wants to know is that its seeing all of them and it shouldn't care about how that is being accomplished.

Now, since the first version, we have started to deviate from that.  I'm not sure if this is still the right thing to do, but I wanted to give a little context.

Reply:
I think the concern about logical/physical separation is only useful if we take CBO into consideration. E.g., for a `AllTuples` distribution requirement, the planner may produce two plans using `SinglePartition` and `BroadcastPartitioning` respectively and pick a cheaper one. In the scope of our current planner framework, this separation doesn't seem to be very useful, though.

Reply:
Sure you can make that argument (I'm not sure I buy it), but then why does this PR still have two different concepts?

Reply:
It seems we should be moving towards CBO, not away from it.

Reply:
I'd like to keep `AllTuples`. `SingleNodeDistribution` is a special case of `AllTuples` and seems we do not really need the extra information introduced by `SingleNode`.

Reply:
I'd like to keep the `Distribution` concept, otherwise it's very weird to say a `RangePartitioning` satisfy `HashPartitioning`, while it looks reasonable to say `RangePartitioning` satisfy `ClusteredDistribution`.

Reply:
Yea, I did miss the point that `ClusteredDistribution` covers both `RangePartitioning` and `HashPartitioning`.